treasuries/src/lib.rs

```diff
@@ -17,47 +17,129 @@
 		type Currency: Currency<Self::AccountId>;
 
 		/// The treasuries' pallet id, used for deriving sovereign account IDs per community.
 		#[pallet::constant]
 		type PalletId: Get<PalletId>;
+
+		// /// the maximum fraction of available treasury funds a single swap can claim
+		// /// defined as divisor: 2 means half of the available funds can be swapped
+		// #[pallet::constant]
+		// type MaxFractionPerSwap: Get<u8>;
+		//
+		// /// the minimum period an account has to wait between two swaps
+		// #[pallet::constant]
+		// type SwapCooldownPeriod: Get<T::Moment>;
+
+		type WeightInfo: WeightInfo;
 	}
 
+	#[pallet::storage]
+	#[pallet::getter(fn swap_native_options)]
+	pub type SwapNativeOptions<T: Config> = StorageDoubleMap<
+		_,
+		Blake2_128Concat,
+		CommunityIdentifier,
+		Blake2_128Concat,
+		T::AccountId,
+		SwapNativeOption<BalanceOf<T>, T::Moment>,
+		OptionQuery,
+	>;
+	#[pallet::call]
+	impl<T: Config> Pallet<T>
+	where
+		sp_core::H256: From<<T as frame_system::Config>::Hash>,
+		T::AccountId: AsRef<[u8; 32]>,
+		u64: TryFrom<BalanceOf<T>>,
+	{
+		/// swap native tokens for community currency subject to an existing swap option for the
+		/// sender account.
+		#[pallet::call_index(0)]
+		#[pallet::weight((<T as Config>::WeightInfo::swap_native(), DispatchClass::Normal, Pays::Yes))]
+		pub fn swap_native(
+			origin: OriginFor<T>,
+			cid: CommunityIdentifier,
+			desired_native_amount: BalanceOf<T>,
+		) -> DispatchResultWithPostInfo {
+			let sender = ensure_signed(origin)?;
+			let swap_option =
+				Self::swap_native_options(cid, &sender).ok_or(<Error<T>>::NoValidSwapOption)?;
+			ensure!(
+				swap_option.native_allowance >= desired_native_amount,
+				Error::<T>::InsufficientAllowance
+			);
+			let treasury_account = Self::get_community_treasury_account_unchecked(Some(cid));
+			ensure!(
+				T::Currency::free_balance(&treasury_account) - T::Currency::minimum_balance() >=
+					desired_native_amount,
+				Error::<T>::InsufficientNativeFunds
+			);
+			let rate = swap_option.rate.ok_or(Error::<T>::SwapRateNotDefined)?;
+			let cc_amount = BalanceType::from_num(
+				u64::try_from(desired_native_amount).or(Err(Error::<T>::SwapOverflow))?,
+			)
+			.checked_mul(rate)
+			.ok_or(Error::<T>::SwapOverflow)?;
+			if swap_option.do_burn {
+				<pallet_encointer_balances::Pallet<T>>::burn(cid, &sender, cc_amount)?;
+			} else {
+				<pallet_encointer_balances::Pallet<T>>::do_transfer(
+					cid,
+					&sender,
+					&treasury_account,
+					cc_amount,
+				)?;
+			}
+			let new_swap_option = SwapNativeOption {
+				native_allowance: swap_option.native_allowance - desired_native_amount,
+				..swap_option
+			};
+			<SwapNativeOptions<T>>::insert(cid, &sender, new_swap_option);
+			Self::do_spend_native(Some(cid), &sender, desired_native_amount)?;
+			Ok(().into())
+		}
+	}
 	impl<T: Config> Pallet<T>
 	where
 		sp_core::H256: From<<T as frame_system::Config>::Hash>,
```

Comment:
Hmm, I actually don't understand why we want to use the `T::Hashing` trait here. We can remove those traitbounds if we directly call the blake256 here.

Additionally, I have a suggestion: https://github.com/encointer/pallets/issues/412

Reply:
using associated types is what we do everywhere else. why would we specify the hasher explicitly in this case? I don't see a problem with it, but it feels unusual

Reply:
I'll leave it as it was already done in faucet. we can refactor this anytime without difficulties

Reply:
Okok, sure. maybe too much attention to the detail here. :)